service/controller/resource/nodecontroller/create.go

```diff
@@ -0,0 +1,156 @@
+		// of controllers takes a non-zero amount of time. By stopping the desired controller within the mutex,
+		// we ensure that only one controller is running at a time and we don't have dangling controllers causing
+		// a memory leak.
+		desired.Stop(ctx)
+	} else {
+		r.controllers[controllerKey] = desired
+	}
+	r.controllerMutex.Unlock()
+
+	r.logger.Debugf(ctx, "controller booted and registered")
+}
+
+// applyUpdateChange checks for an existing controller, stops it if it doesn't match the desired controller,
+// and returns a value indicating if the desired controller should be started (because the existing one was
+// stopped or there was no existing one).
+func (r *Resource) applyUpdateChange(ctx context.Context, desired controllerWithConfig) bool {
+	r.controllerMutex.Lock()         // Exclude access from other goroutines
+	defer r.controllerMutex.Unlock() // Ensure this is run for all function returns
+
+	controllerKey := controllerMapKey(desired.cluster)
+	current, ok := r.controllers[controllerKey]
+
+	if !ok {
+		r.logger.Debugf(ctx, "existing controller not found")
+		return true
+	}
+
+	if !r.controllersEqual(ctx, current, desired) {
+		r.logger.Debugf(ctx, "existing controller needs to be replaced, stopping and deleting")
+		current.Stop(ctx) // This is just closing a channel, should be almost instant
+		delete(r.controllers, controllerKey)
+		return true
+	}
+
+	r.logger.Debugf(ctx, "existing controller doesn't need to be replaced")
+	return false
+}
+
+// calculateDesiredController returns a controllerWithConfig holding a controller which watches the workload cluster
+// and other information used to determine when the controller needs to be created or updated.
+func (r *Resource) calculateDesiredController(ctx context.Context, cluster v1alpha1.KVMConfig) (controllerWithConfig, error) {
+	restConfig, err := r.workloadCluster.NewRestConfig(ctx, key.ClusterID(cluster), key.ClusterAPIEndpoint(cluster))
+	if err != nil {
+		return controllerWithConfig{}, microerror.Mask(err)
+	}
+
+	k8sClient, err := k8sclient.NewClients(k8sclient.ClientsConfig{
+		Logger:     r.logger,
+		RestConfig: restConfig,
+	})
+	if err != nil {
+		return controllerWithConfig{}, microerror.Mask(err)
+	}
+
+	nodeController, err := nodecontroller.New(nodecontroller.Config{
+		Cluster:             cluster,
+		ManagementK8sClient: r.k8sClient,
+		WorkloadK8sClient:   k8sClient,
+		Logger:              r.logger,
+	})
+	if err != nil {
+		return controllerWithConfig{}, microerror.Mask(err)
+	}
+
+	withConfig := controllerWithConfig{
+		cluster:    cluster,
+		restConfig: restConfig,
+		Controller: nodeController,
+	}
+	return withConfig, nil
+}
+
+// controllersEqual returns true when the given controllers are equal as it relates to watching the workload
+// cluster Kubernetes APIs.
+func (r *Resource) controllersEqual(ctx context.Context, left controllerWithConfig, right controllerWithConfig) bool {
```

Comment:
I'm wondering if this logic (and its helpers) should belong to `controllerWithConfig`, either statically or as an instance method (e.g. `left.Equals(right)`). That way if its API changes this implementation is more likely to follow.

Reply:
I agree that's it's better to keep them close. I made it into a method on the resource handler so we could have the logger, but those lines are just for my debugging so I'll drop them and move the function to be a member of `controllerWithConfig`.

Reply:
Fixed.